llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8398,19 +8340,84 @@ const Vkgc::PipelineOptions *SPIRVToLLVM::getPipelineOptions() const {
 }
 
-// This method does preparation work for adding the scratch access bounds checks for memory operations
-// like OpLoads and OpStores. We are handling private and function memory classes here.
-// The idea behind this method is that while translating a OpLoad or OpStore which loads from local scratch memory using
-// a unknown (run-time) index into an array, we parse the operands of the corresponding access chain, add a chain of
-// ANDed compares to check if the bounds requirements are met and then do a conditional jump to either the intermediate
-// block which handles the memory operation or the final block (which will skip the memory op, if bounds checks failed).
-// As mentioned, this does preparation work like inserting the bounds checks, its up to the caller caring about
-// inserting the OpLoad or OpStore. This also sets a flag that signals the main translation loop that we want to
-// continue inserting subsequent instructions at the final block instead of just going back to the parent block.
-// Currently, OpLoads will yield zero and OpStores will just omit the store, if bounds checks failed. This is up to the
-// caller. This implementation handles nested structs using arrays and 1...n-dimensional arrays. This returns if we need
-// to insert a bounds checks and pointers to the newly created basic blocks (which can be optimized away by subsequent
-// passes, if not needed).
+bool SPIRVToLLVM::shouldInsertScratchBoundsCheck(SPIRVValue *accessChain) const {
+  if (!getPipelineOptions()->enableScratchAccessBoundsChecks) {
+    return false;
+  }
+
+  if (!accessChain || accessChain->getOpCode() != OpAccessChain) {
+    return false;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = accessChain->getType()->getPointerStorageClass();
+
+  // We don't want bounds checks if we are outside private or function storage class.
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return false;
+  }
+
+  return true;
+}
+
+Value *SPIRVToLLVM::getLastInsertedValue() {
+  BasicBlock *insertBlock = getBuilder()->GetInsertBlock();
+  if (!insertBlock->getInstList().empty()) {
+    return &insertBlock->getInstList().back();
+  }
+
+  if (insertBlock->empty()) {
+    return nullptr;
+  }
+
+  return insertBlock;
+}
+
+// Records all instructions from baseNode to the latest inserted instructions and stores the instruction classification
+// here. If we have used addLoadInstRecursively, these will be loads, if we have used addStoreInstRecursively, these
+// will be stores. In the load case, the last instruction will be connected with the subsequent instructions by
+// inserting a phi node.
+void SPIRVToLLVM::gatherScratchBoundsChecksMemops(SPIRVValue *accessChain, Value *baseNode,
+                                                  SPIRVToLLVM::LlvmMemopType instructionType, SPIRVValue *valueMapKey) {
+  if (!shouldInsertScratchBoundsCheck(accessChain) || !baseNode) {
```

Comment:
This function should run normally if `baseNode` is `nullptr`. Otherwise it won’t insert a bounds check if a load is the first instruction in a basic block.

Reply:
This won't work if ```getLastInsertedValue``` returns a ```nullptr```, because the loop won't be executed. If the BB is empty, there is no need to insert a bounds check. That's what the ```return insertBlock;``` was used for: record all instructions from the beginning to anything after the ```add...``` call in case no instruction was inserted before (e. g., if the BB was empty, just use an iterator to its begin) and record all instructions.

Reply:
Oh, I see. Somehow we need to make that case work. I think the line below `baseNode ? dyn_cast<Instruction>(baseNode) : &*(dyn_cast<BasicBlock>(baseNode)->begin());` needs to be changed for that. (I think the second case in there would only be hit if `baseNode` is null?)

Reply:
I think, passing the according instruction itself should eliminate the need for the cast here and also fix the issue when the load is the first instruction in a BB:

```
if (baseNode) {
gather [baseNode, insertPos];
} else {
gather [BB->begin(), insertPos]; // could be equal
}
```

Reply:
Sounds good

Reply:
Added a change for this part.